Sources/Fuzzilli/FuzzIL/AbstractInterpreter.swift

```diff
@@ -11,17 +11,108 @@
-public struct AbstractInterpreter {
-    // States are kept in a stack to support conditional execution.
+private struct InterpreterState {
+    // State diffs are kept in a stacks to support conditional execution.
+    // Save final variable types changed in if block.
     private var ifStack: [VariableMap<Type>] = []
-    private var stack = [VariableMap<Type>()]
+    // Save last variable type before changing it in current block.
+    // It is used to reset/merge previous states to current one.
+    private var stateDiffsStack = [VariableMap<Type>()]
     
     // The currently active state.
-    private var currentState: VariableMap<Type> {
-        return stack.last!
+    private var currentState = VariableMap<Type>()
+
+    public mutating func reset() {
+        ifStack.removeAll()
+        stateDiffsStack = [VariableMap<Type>()]
+        currentState = VariableMap<Type>()
+    }
+
+    public enum StackType {
+        case conditional
+        case main
+    }
+
+    /// Update current variable types and type diffs
+    /// Used after block end when some state is removed
+    public mutating func mergeStates(
+        from stackType: StackType = .main, typeChanges: inout [(Variable, Type)], reset: Bool = false
+    ) {
+        let lastStateDiff: VariableMap<Type>
+        switch stackType {
+        case .main:
+            lastStateDiff = stateDiffsStack.removeLast()
+        case .conditional:
+            lastStateDiff = ifStack.removeLast()
+        }
+
+        for (variable, oldType) in lastStateDiff {
+            var newType = oldType
+
+            if !reset {
+                // .nothing cannot be unioned with other type
+                if oldType != .nothing {
+                    newType |= currentState[variable]!
+                }
+            }
+
+            if currentState[variable] != newType {
+                // .nothing means variable no longer exists
+                // This information does not need to be forwarded in typeChanges
+                if newType == .nothing {
+                    currentState[variable] = nil
+                } else {
+                    currentState[variable] = newType
+                    typeChanges.append((variable, newType))
+                }
+            }
+
+            // We should also update variable type diff if there was none in prev state
+            if stateDiffsStack[stateDiffsStack.count - 1][variable] == nil {
+                stateDiffsStack[stateDiffsStack.count - 1][variable] = oldType
+            }
+        }
+    }
+
+    // Save final variable types in if block
+    public mutating func saveIfState() {
+        ifStack.append(VariableMap())
+        // Iterate over variables that were changed in last if block
+        for (variable, oldType) in stateDiffsStack.last! {
+            // Save only non-local variables
+            if oldType != .nothing {
+                ifStack[ifStack.count - 1][variable] = currentState[variable]
```

Comment:
Nit: maybe it would be nicer to keep the `VariableMap` in a local variable first, then append it to the stack after this loop?